cyto_dl/models/contrastive/contrastive.py

```diff
@@ -0,0 +1,141 @@
+from pathlib import Path
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import torch
+import torch.nn as nn
+from sklearn.decomposition import PCA
+from torchmetrics import MeanMetric
+
+from cyto_dl.models.base_model import BaseModel
+
+
+class Contrastive(BaseModel):
+    def __init__(
+        self,
+        backbone: nn.Module,
+        task_head: nn.Module,
+        anchor_key: str = "image",
+        positive_key: str = "image_aug",
+        target_key: str = "target",
+        meta_keys: list[str] = [],
+        save_dir: str = "./",
+        viz_freq: int = 10,
+        **base_kwargs,
+    ):
+        """
+        Parameters
+        ----------
+        backbone: nn.Module
+            Backbone model
+        task_head: nn.Module
+            Task head model
+        anchor_key: str
+            Key in batch dictionary for anchor image
+        positive_key: str
+            Key in batch dictionary for positive image
+        target_key: str
+            OPTIONAL Key in batch dictionary for target, used only for visualization
+        meta_keys: list[str]
+            List of keys in batch dictionary to save to csv during prediction
+        save_dir: str
+            Directory to save visualizations
+        viz_freq: int
+            Frequency to save visualizations
+        **base_kwargs:
+            Additional arguments passed to BaseModel
+        """
+        _DEFAULT_METRICS = {
+            "train/loss": MeanMetric(),
+            "val/loss": MeanMetric(),
+            "test/loss": MeanMetric(),
+        }
+        metrics = base_kwargs.pop("metrics", _DEFAULT_METRICS)
+        super().__init__(metrics=metrics, **base_kwargs)
+
+        self.backbone = backbone
+        self.task_head = task_head
+
+    def forward(self, x1, x2):
+        return self.backbone(x1), self.backbone(x2)
+
+    def plot_neighbors(self, embedding1, embedding2):
```

Comment:
these are callbacks?

Reply:
I'm kind of inclined to leave them as functions unless they seem generally useful

Reply:
cool